exercises/binary-search/README.md

```diff
@@ -0,0 +1,113 @@
+# Binary Search
+
+Write a program that implements a binary search algorithm.
+
+Searching a sorted collection is a common task. A dictionary is a sorted
+list of word definitions. Given a word, one can find its definition. A
+telephone book is a sorted list of people's names, addresses, and
+telephone numbers. Knowing someone's name allows one to quickly find
+their telephone number and address.
+
+If the list to be searched contains more than a few items (a dozen, say)
+a binary search will require far fewer comparisons than a linear search,
+but it imposes the requirement that the list be sorted.
+
+In computer science, a binary search or half-interval search algorithm
+finds the position of a specified input value (the search "key") within
+an array sorted by key value.
+
+In each step, the algorithm compares the search key value with the key
+value of the middle element of the array.
+
+If the keys match, then a matching element has been found and its index,
+or position, is returned.
+
+Otherwise, if the search key is less than the middle element's key, then
+the algorithm repeats its action on the sub-array to the left of the
+middle element or, if the search key is greater, on the sub-array to the
+right.
+
+If the remaining array to be searched is empty, then the key cannot be
+found in the array and a special "not found" indication is returned.
+
+A binary search halves the number of items to check with each iteration,
+so locating an item (or determining its absence) takes logarithmic time.
+A binary search is a dichotomic divide and conquer search algorithm.
+
+## Restrictions
```

Comment:
This text is going to have go in hints.md, otherwise it disappears when README is regenerated.

Reply:
Do you mean the *Restrictions* section? Actually it's not really a hint but a part of the exercise. So I guess it should go into the main exercise text?

Reply:
In https://github.com/exercism/rust/blob/master/exercises/decimal/README.md a "# note" section is used for similar advice. Is this the right section to put it into?

Reply:
> Do you mean the *Restrictions* section?

I should have been more clear, therefore I apologise for not being so.

My goal is that the README given in this PR is exactly the README generated by `configlet generate` (documented at https://github.com/exercism/docs/blob/master/maintaining-a-track/regenerating-exercise-readmes.md), with no modifications. Because of time, the track maintainers are unable to maintain any README that is different from the generated README.

So, the entire section needs to go into hints.md, otherwise it will disappear.

I will be generating the README, so you do not need to fix it yourself, but I need you to know it for future PRs.